airbyte-integrations/connectors/source-trustpilot/source_trustpilot/source.py

```diff
@@ -0,0 +1,178 @@
+#
+import datetime
+import json
+from dataclasses import InitVar, dataclass, field
+from html.parser import HTMLParser
+from time import sleep
+from typing import Any, Mapping, Optional, Union, Iterable
+from typing import List
+
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.datetime import MinMaxDatetime
+from airbyte_cdk.sources.declarative.datetime.datetime_parser import DatetimeParser
+from airbyte_cdk.sources.declarative.extractors.record_extractor import RecordExtractor
+from airbyte_cdk.sources.declarative.interpolation import InterpolatedString
+from airbyte_cdk.sources.declarative.requesters.paginators import DefaultPaginator
+from airbyte_cdk.sources.declarative.requesters.request_option import RequestOptionType
+from airbyte_cdk.sources.declarative.stream_slicers import SingleSlice
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState, Config
+from airbyte_cdk.sources.declarative.yaml_declarative_source import YamlDeclarativeSource
+
+
+class TrustpilotReviewsParser(HTMLParser):
+    def __init__(self):
+        super().__init__()
+
+        self.__inside_reviews_script = False
+        self.__reviews_script_data = None
+
+    def handle_starttag(self, tag, attrs):
+        if tag == 'script':
+            attrs_map = {e[0]: e[1] for e in attrs}
+            if (('type' in attrs_map) and (attrs_map['type'] == 'application/ld+json')) and \
+                    (('data-business-unit-json-ld' in attrs_map) and (attrs_map['data-business-unit-json-ld'] == 'true')):
+                self.__inside_reviews_script = True
+
+    def handle_data(self, data):
+        if self.__inside_reviews_script:
+            self.__reviews_script_data = data
+
+    def handle_endtag(self, tag):
+        if tag == 'script':
+            if self.__inside_reviews_script:
+                self.__inside_reviews_script = False
+
+    def get_reviews_script_data(self):
+        return self.__reviews_script_data
+
+
+@dataclass
+class TrustpilotExtractor(RecordExtractor):
+    """
+    Extractor for Trustpilot.
+    It parses output HTML and finds all `Review` occurences in it.
+    """
+
+    options: InitVar[Mapping[str, Any]]
+
+    def __post_init__(self, options: Mapping[str, Any]):
+        self._timeout_ms = options['config']['timeout_ms'] if 'timeout_ms' in options['config'] else 1000
+
+    @staticmethod
+    def __parse_html(response: requests.Response) -> List[Record]:
+        dom = response.text
+        parser = TrustpilotReviewsParser()
+        parser.feed(dom)
+        data = parser.get_reviews_script_data()
+        data = json.loads(data)
+        data = data['@graph']
+        data = list(filter(lambda e: e.get('@type', '') == 'Review', data))
+        return data
+
+    def extract_records(self, response: requests.Response,
+                        ) -> List[Record]:
+        sleep(self._timeout_ms / 1000.0)
+        return self.__parse_html(response)
```

Comment:
Please add unit testing for all these functions and class. You don't need to use a complete html reponse but use one to help future maintaince.

Reply:
Okay, great suggestion!
Thank You for advice!

Reply:
Did it.